snare/tests/test_get_body.py

```diff
@@ -0,0 +1,120 @@
+        self.filename, self.hashname = self.handler._make_filename(yarl.URL(self.root))
+        self.meta = {'/index.html': {'content_type': 'text/html', 'hash': 'd1546d731a9f30cc80127d57142a482b'},
+                     '/test': {'content_type': 'text/html', 'hash': '4539330648b80f94ef3bf911f6d77ac9'}}
+
+        async def test():
+            await self.handler.new_urls.put((yarl.URL(self.root), 0))
+            await self.handler.get_body(self.session)
+
+        self.loop.run_until_complete(test())
+        with open(os.path.join(self.target_path, self.hashname)) as f:
+            self.return_content = f.read()
+
+        self.expected_content = '<html><body><a href="/test"></a></body></html>'
+        self.assertEqual(self.return_content, self.expected_content)
+
+        self.assertEqual(self.handler.visited_urls[-1], 'http://example.com/test')
+        self.assertEqual(self.handler.meta, self.meta)
+
+    def test_get_body_css_validate(self):
+        aiohttp.ClientResponse._headers = {'Content-Type': 'text/css'}
+
+        self.css_validate = 'true'
+        self.handler = Cloner(self.root, self.max_depth, self.css_validate)
+        self.content = b'''.banner { background: url("/example.png") }'''
+        aiohttp.ClientResponse.read = AsyncMock(return_value=self.content)
+
+        async def test():
+            await self.handler.new_urls.put((yarl.URL(self.root), 0))
+            self.return_content = await self.handler.get_body(self.session)
+
+        self.loop.run_until_complete(test())
+        self.expected_content = yarl.URL('http://example.com/example.png')
+        self.assertEqual(self.return_content, self.expected_content)
+
+        async def get_url():
+            self.url, self.level = await self.handler.new_urls.get()
+
+        self.loop.run_until_complete(get_url())
+        self.assertEqual(self.url, self.expected_content)
+        self.assertEqual(self.level, 1)
+
+    def test_get_body_css_validate_scheme(self):
+        aiohttp.ClientResponse._headers = {'Content-Type': 'text/css'}
+
+        self.css_validate = 'true'
+        self.handler = Cloner(self.root, self.max_depth, self.css_validate)
+        self.content = b'''.banner { background: url("data://domain/test.txt") }'''
+        self.expected_content = None
+        aiohttp.ClientResponse.read = AsyncMock(return_value=self.content)
+
+        async def test():
+            await self.handler.new_urls.put((yarl.URL(self.root), 0))
+            self.return_content = await self.handler.get_body(self.session)
+
+        self.loop.run_until_complete(test())
+        self.assertEqual(self.return_content, self.expected_content)
+
+        self.content = b'''.banner { background: url("file://domain/test.txt") }'''
+        aiohttp.ClientResponse.read = AsyncMock(return_value=self.content)
+        self.loop.run_until_complete(test())
+        self.assertEqual(self.return_content, self.expected_content)
```

Comment:
Writing it in a loop will avoid code repetition. Also, check no changes in changes in global variables.